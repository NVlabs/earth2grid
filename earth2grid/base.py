# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from typing import Protocol

import numpy as np


class Grid(Protocol):
    """lat and lon should be broadcastable arrays"""

    @property
    def lat(self):
        pass

    @property
    def lon(self):
        pass

    @property
    def shape(self) -> tuple[int, ...]:
        pass

    def get_bilinear_regridder_to(self, lat: np.ndarray, lon: np.ndarray):
        """Return a regridder from `self` to lat/lon.

        Args:
            lat, lon: broadcastable arrays for the lat/lon
        """
        raise NotImplementedError()

    def visualize(self, data):
        pass

    def to_pyvista(self):
        pass
