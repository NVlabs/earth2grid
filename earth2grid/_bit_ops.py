# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


def compact_bits(bits):
    # Remove interleaved 0 bits
    bits = bits & 0x5555555555555555  # Mask: 01010101...
    # example implementation for 1 byte
    # 0a0b0c0d
    # 00ab00cd # (x | x >> 1)  & 00110011 = 0x33
    # 0000abcd # (x | x >> 2)  & 00001111 = 0x0F
    # --------
    # abc0d
    bits = (bits | (bits >> 1)) & 0x3333333333333333  # noqa
    bits = (bits | (bits >> 2)) & 0x0F0F0F0F0F0F0F0F  # noqa
    bits = (bits | (bits >> 4)) & 0x00FF00FF00FF00FF  # noqa
    bits = (bits | (bits >> 8)) & 0x0000FFFF0000FFFF  # noqa
    bits = (bits | (bits >> 16)) & 0x00000000FFFFFFFF  # noqa
    return bits


def spread_bits(bits):
    """
    bits is a 32 bit number (stored in int64)
    algorithm starts by moving the first 16 bits to the left by 16
    and proceeding recursively
    """
    # example implementation for a byte
    # 0000abcd
    # 00ab00cd # (x | x <<2)  & 00110011 = 0x33
    # 0a0b0c0d # (x | x <<1)  & 01010100 = 0x55
    # --------
    # abc0d
    bits = (bits | (bits << 16)) & 0x0000FFFF0000FFFF  # noqa
    bits = (bits | (bits << 8)) & 0x00FF00FF00FF00FF  # noqa
    bits = (bits | (bits << 4)) & 0x0F0F0F0F0F0F0F0F  # noqa
    bits = (bits | (bits << 2)) & 0x3333333333333333  # noqa
    bits = (bits | (bits << 1)) & 0x5555555555555555  # noqa
    return bits
