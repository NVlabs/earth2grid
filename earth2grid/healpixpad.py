# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License
#
# Written by Mauro Bisson <maurob@nvidia.com> and Thorsten Kurth <tkurth@nvidia.com>.


import healpixpad_cuda
import torch


class HEALPixPadFunction(torch.autograd.Function):
    """
    A torch autograd class that pads a healpixpad xy tensor
    """
    @staticmethod
    def forward(ctx, input, pad):
        """
        The forward pass of the padding class

        Parameters
        ----------
        input: torch.tensor
            The tensor to pad, must have 5 dimensions and be in (B, F, C, H, W) format
            where F == 12 and H == W
        pad: int
            The amount to pad each face of the tensor

        Returns
        -------
        torch.tensor: The padded tensor
        """
        ctx.pad = pad
        if len(input.shape) != 5:
            raise ValueError(f"Input tensor must be have 4 dimensions (B, F, C, H, W), got {len(input.shape)} dimensions instead")
        # make contiguous
        input = input.contiguous()
        out = healpixpad_cuda.forward(input, pad)[0]
        return out

    @staticmethod
    def backward(ctx, grad):
        pad = ctx.pad
        grad = grad.contiguous()
        out = healpixpad_cuda.backward(grad, pad)[0]
        return out, None

class HEALPixPad(torch.nn.Module):
    """
    A torch module that handles padding of healpixpad xy tensors

    Paramaeters
    -----------
    padding: int
        The amount to pad the tensors
    """
    def __init__(self, padding: int):
        super(HEALPixPad, self).__init__()
        self.padding = padding

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        """
        The forward pass of the padding class

        Parameters
        ----------
        input: torch.tensor
            The tensor to pad, must have 5 dimensions and be in (B, F, C, H, W) format
            where F == 12 and H == W

        Returns
        -------
        torch.tensor: The padded tensor
        """
        return HEALPixPadFunction.apply(input, self.padding)

