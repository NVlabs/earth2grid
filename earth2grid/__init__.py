# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import torch

from earth2grid import base, healpix, latlon
from earth2grid._regrid import (
    BilinearInterpolator,
    Identity,
    S2NearestNeighborInterpolator,
)

__all__ = [
    "base",
    "healpix",
    "latlon",
    "get_regridder",
    "BilinearInterpolator",
    "S2NearestNeighborInterpolator",
    "S2LinearBarycentricInterpolator",
]


def get_regridder(src: base.Grid, dest: base.Grid) -> torch.nn.Module:
    """Get a regridder from `src` to `dest`"""
    if src == dest:
        return Identity()
    elif isinstance(src, latlon.LatLonGrid) and isinstance(dest, latlon.LatLonGrid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(src, latlon.LatLonGrid) and isinstance(dest, healpix.Grid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(src, healpix.Grid):
        return src.get_bilinear_regridder_to(dest.lat, dest.lon)
    elif isinstance(dest, healpix.Grid):
        return src.get_healpix_regridder(dest)  # type: ignore

    raise ValueError(src, dest, "not supported.")
