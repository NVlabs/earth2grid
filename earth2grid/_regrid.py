# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import einops
import netCDF4 as nc
import numpy as np
import pandas
import torch

from earth2grid import base, healpix
from earth2grid.latlon import LatLonGrid


class BilinearInterpolator(torch.nn.Module):
    """Bilinear interpolation for a non-uniform grid"""

    def __init__(
        self, x_coords: torch.Tensor, y_coords: torch.Tensor, x_query: torch.Tensor, y_query: torch.Tensor
    ) -> None:
        """

        Args:
            x_coords (Tensor): X-coordinates of the input grid, shape [W]. Must be in increasing sorted order.
            y_coords (Tensor): Y-coordinates of the input grid, shape [H]. Must be in increasing sorted order.
            x_query (Tensor): X-coordinates for query points, shape [N].
            y_query (Tensor): Y-coordinates for query points, shape [N].
        """
        super().__init__()

        # Ensure input coordinates are float for interpolation
        x_coords, y_coords = x_coords.float(), y_coords.float()

        if torch.any(x_coords[1:] <= x_coords[:-1]):
            raise ValueError("x_coords must be in increasing order.")

        if torch.any(y_coords[1:] <= y_coords[:-1]):
            raise ValueError("y_coords must be in increasing order.")

        # Find indices for the closest lower and upper bounds in x and y directions
        x_l_idx = torch.searchsorted(x_coords, x_query, right=True) - 1
        x_u_idx = x_l_idx + 1
        y_l_idx = torch.searchsorted(y_coords, y_query, right=True) - 1
        y_u_idx = y_l_idx + 1

        # Clip indices to ensure they are within the bounds of the input grid
        x_l_idx = x_l_idx.clamp(0, x_coords.size(0) - 2)
        x_u_idx = x_u_idx.clamp(1, x_coords.size(0) - 1)
        y_l_idx = y_l_idx.clamp(0, y_coords.size(0) - 2)
        y_u_idx = y_u_idx.clamp(1, y_coords.size(0) - 1)

        # Compute weights
        x_l_weight = (x_coords[x_u_idx] - x_query) / (x_coords[x_u_idx] - x_coords[x_l_idx])
        x_u_weight = (x_query - x_coords[x_l_idx]) / (x_coords[x_u_idx] - x_coords[x_l_idx])
        y_l_weight = (y_coords[y_u_idx] - y_query) / (y_coords[y_u_idx] - y_coords[y_l_idx])
        y_u_weight = (y_query - y_coords[y_l_idx]) / (y_coords[y_u_idx] - y_coords[y_l_idx])

        self.x_l_idx = x_l_idx
        self.x_u_idx = x_u_idx
        self.y_l_idx = y_l_idx
        self.y_u_idx = y_u_idx

        self.register_buffer("x_l_weight", x_l_weight)
        self.register_buffer("x_u_weight", x_u_weight)
        self.register_buffer("y_l_weight", y_l_weight)
        self.register_buffer("y_u_weight", y_u_weight)

    def forward(self, z: torch.Tensor):
        """
        Interpolate the field

        Args:
            z: shape [Y, X]
        """

        # Perform bilinear interpolation
        interpolated_values = (
            z[..., self.y_l_idx, self.x_l_idx] * self.x_l_weight * self.y_l_weight
            + z[..., self.y_l_idx, self.x_u_idx] * self.x_u_weight * self.y_l_weight
            + z[..., self.y_u_idx, self.x_l_idx] * self.x_l_weight * self.y_u_weight
            + z[..., self.y_u_idx, self.x_u_idx] * self.x_u_weight * self.y_u_weight
        )
        return interpolated_values


class TempestRegridder(torch.nn.Module):
    def __init__(self, file_path):
        super().__init__()
        dataset = nc.Dataset(file_path)
        self.lat = dataset["latc_b"][:]
        self.lon = dataset["lonc_b"][:]

        i = dataset["row"][:] - 1
        j = dataset["col"][:] - 1
        M = dataset["S"][:]

        i = i.data
        j = j.data
        M = M.data

        self.M = torch.sparse_coo_tensor((i, j), M, [max(i) + 1, max(j) + 1]).float()

    def to(self, device):
        self.M = self.M.to(device)
        return self

    def forward(self, x):
        xr = einops.rearrange(x, "b c x y -> b c (x y)")
        yr = xr @ self.M.T
        y = einops.rearrange(yr, "b c (x y) -> b c x y", x=self.lat.size, y=self.lon.size)
        return y


class Identity(torch.nn.Module):
    def forward(self, x):
        return x


class RegridLatLon(torch.nn.Module):
    def __init__(self, src_grid: LatLonGrid, dest_grid: LatLonGrid):
        super().__init__()
        self._src_grid = src_grid
        self._dest_grid = dest_grid
        self._lat_index = pandas.Index(src_grid.lat).get_indexer(dest_grid.lat)
        assert not np.any(self._lat_index == -1)  # noqa

        self._lon_index = pandas.Index(src_grid.lon).get_indexer(dest_grid.lon)
        assert not np.any(self._lon_index == -1)  # noqa

    def forward(self, x):
        if x.shape[-2:] != self._src_grid.shape:
            raise ValueError(f"Input shape {x.shape} does not match grid shape" f"{self._src_grid.shape}")

        return x[..., self._lat_index, :][..., self._lon_index]


class _RegridFromLatLon(torch.nn.Module):
    """Regrid from lat-lon to unstructured grid with bilinear interpolation"""

    def __init__(self, src: LatLonGrid, dest: base.Grid):
        super().__init__()

        # potentially relax this
        assert len(dest.shape) == 1

        # TODO add device switching logic (maybe use torch registers for this
        # info)
        long = np.concatenate([src.lon.ravel(), [360]], axis=-1)
        long_t = torch.from_numpy(long)

        # flip the order latg since bilinear only works with increasing coordinate values
        lat_increasing = src.lat[1] > src.lat[0]
        latg_t = torch.from_numpy(src.lat.ravel())
        lat_query = torch.from_numpy(dest.lat.ravel())

        if not lat_increasing:
            lat_query = -lat_query
            latg_t = -latg_t

        self._bilinear = BilinearInterpolator(
            long_t, latg_t, y_query=lat_query, x_query=torch.from_numpy(dest.lon.ravel())
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # pad z in lon direction
        # only works for a global grid
        # TODO generalize this to local grids and add options for padding
        x = torch.cat([x, x[..., 0:1]], axis=-1)
        return self._bilinear(x)


def get_regridder(src: base.Grid, dest: base.Grid) -> torch.nn.Module:
    """Get a regridder from `src` to `dest`"""
    if src == dest:
        return Identity()
    elif isinstance(src, LatLonGrid) and isinstance(dest, LatLonGrid):
        return RegridLatLon(src, dest)
    elif isinstance(src, LatLonGrid) and isinstance(dest, healpix.Grid):
        return _RegridFromLatLon(src, dest)
    elif isinstance(src, healpix.Grid) and isinstance(dest, LatLonGrid):
        return src.get_latlon_regridder(dest.lat, dest.lon)
    raise ValueError(src, dest, "not supported.")
