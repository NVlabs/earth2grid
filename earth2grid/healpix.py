"""

From this notebook: https://colab.research.google.com/drive/1MzTyeNFiy-7RNY6UtGKsmDavX5dk6epU


Healpy has two indexing conventions NEST and RING. But for convolutions we want
2D array indexing in row or column major order. Here are some vectorized
routines `nest2xy` and `x2nest` for going in between these conventions. The
previous code shared by Dale used string computations to handle these
operations, which was probably quite slow. Here we use vectorized bit-shifting.
"""
from dataclasses import dataclass
from enum import Enum

import einops
import healpy
import numpy as np

try:
    import pyvista as pv
except ImportError:
    pv = None

from earth2grid import base


class PixelOrder(Enum):
    RING = 0
    NEST = 1
    XY = 2


@dataclass
class Grid(base.Grid):
    """A Healpix Grid"""
    level: int
    pixel_order: PixelOrder = PixelOrder.RING

    def _nside(self):
        return 2**self.level

    def _npix(self):
        return self._nside() ** 2 * 12

    def _nest_ipix(self):
        """convert to nested index number"""
        i = np.arange(self._npix())
        if self.pixel_order == PixelOrder.RING:
            return healpy.ring2nest(self._nside(), i)
        elif self.pixel_order == PixelOrder.NEST:
            return i
        else:
            return xy2nest(self._nside(), i)

    def _reorder_to_nest(self, map):
        """convert to nested index number"""
        i = np.arange(self._npix())
        if self.pixel_order == PixelOrder.RING:
            j = healpy.nest2ring(self._nside(), i)
            return map[j]
        elif self.pixel_order == PixelOrder.NEST:
            return map
        else:
            j = nest2xy(self._nside(), i)
            return map[j]

    @property
    def lat(self):
        _, lat = healpy.pix2ang(self._nside(), self._nest_ipix(), lonlat=True, nest=True)
        return lat

    @property
    def lon(self):
        lon, _ = healpy.pix2ang(self._nside(), self._nest_ipix(), lonlat=True, nest=True)
        return lon

    @property
    def shape(self) -> tuple[int, ...]:
        return (self._npix(),)

    def visualize(self, map):
        healpy.mollview(self._reorder_to_nest(map), nest=True)

    def to_pyvista(self):
        if pv is None:
            raise ImportError("Need to install pyvista")

        # Make grid
        nside = 2**self.level
        pix = self._nest_ipix()
        points = healpy.boundaries(nside, pix, step=1, nest=True)
        out = einops.rearrange(points, "n d s -> (n s) d")
        unique_points, inverse = np.unique(out, return_inverse=True, axis=0)
        assert unique_points.ndim == 2
        assert unique_points.shape[1] == 3
        inverse = einops.rearrange(inverse, "(n s) -> n s", n=pix.size)
        n, s = inverse.shape
        cells = np.ones_like(inverse, shape=(n, s + 1))
        cells[:, 0] = s
        cells[:, 1:] = inverse
        celltypes = np.full(shape=(n,), fill_value=pv.CellType.QUAD)
        grid = pv.UnstructuredGrid(cells, celltypes, unique_points)
        return grid


# nside = 2^ZOOM_LEVELS
ZOOM_LEVELS = 12


def _extract_every_other_bit(binary_number):
    result = 0
    shift_count = 0

    for i in range(ZOOM_LEVELS):
        # Check if the least significant bit is 1
        # Set the corresponding bit in the result
        result |= (binary_number & 1) << shift_count

        # Shift to the next bit to check
        binary_number = binary_number >> 2
        shift_count += 1

    return result


def nest2xy(nside, i):
    """convert NEST to XY index"""
    tile = i // nside**2
    j = i % (nside**2)
    x = _extract_every_other_bit(j)
    y = _extract_every_other_bit(j >> 1)
    return tile * nside**2 + y * nside + x


def xy2nest(nside, i):
    """convert XY index to NEST"""
    tile = i // (nside**2)
    y = (i % (nside**2)) // nside
    x = i % nside

    result = 0
    for i in range(ZOOM_LEVELS):
        # Extract the ith bit from the number
        extracted_bit = (x >> i) & 1
        result |= extracted_bit << (2 * i)

        extracted_bit = (y >> i) & 1
        result |= extracted_bit << (2 * i + 1)
    return result | (tile * nside**2)
