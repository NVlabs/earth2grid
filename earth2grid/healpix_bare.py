# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import torch

from earth2grid import _healpix_bare
from earth2grid._healpix_bare import corners, hpc2loc, hpd2loc, nest2hpd, nest2ring, ring2hpd, ring2nest

__all__ = [
    "pix2ang",
    "ring2nest",
    "nest2ring",
    "hpc2loc",
    "corners",
]


def pix2ang(nside, i, nest=False, lonlat=False):
    if nest:
        hpd = nest2hpd(nside, i)
    else:
        hpd = ring2hpd(nside, i)
    loc = hpd2loc(nside, hpd)
    lon, lat = _loc2ang(loc)

    if lonlat:
        return torch.rad2deg(lon), 90 - torch.rad2deg(lat)
    else:
        return lat, lon


def _loc2ang(loc):
    """
    static t_ang loc2ang(tloc loc)
    { return (t_ang){atan2(loc.s,loc.z), loc.phi}; }
    """
    z = loc[..., 0]
    s = loc[..., 1]
    phi = loc[..., 2]
    return phi % (2 * torch.pi), torch.atan2(s, z)


def loc2vec(loc):
    z = loc[..., 0]
    s = loc[..., 1]
    phi = loc[..., 2]
    x = (s * torch.cos(phi),)
    y = (s * torch.sin(phi),)
    return x, y, z


def get_interp_weights(nside: int, lon: torch.Tensor, lat: torch.Tensor):
    """

    Args:
        lon: longtiude in deg E. Shape (*)
        lat: latitdue in deg E. Shape (*)

    Returns:
        pix, weights: both shaped (4, *). pix is given in RING convention.

    """
    shape = lon.shape
    lon = lon.double().cpu().flatten()
    lat = lat.double().cpu().flatten()
    pix, weights = _healpix_bare.get_interp_weights(nside, lon, lat)

    return pix.view(4, *shape), weights.view(4, *shape)
