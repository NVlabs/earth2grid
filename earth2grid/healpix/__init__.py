# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""

HEALPix
=======

From this notebook: https://colab.research.google.com/drive/1MzTyeNFiy-7RNY6UtGKsmDavX5dk6epU



"""


from earth2grid.healpix._padding import PaddingBackends, pad, pad_backend, pad_with_dim
from earth2grid.healpix.coordinates import (
    angular_to_global,
    global_to_angular,
    global_to_face,
)
from earth2grid.healpix.core import (
    HEALPIX_PAD_XY,
    XY,
    Compass,
    Grid,
    PixelOrder,
    local2xy,
    nest2xy,
    npix2level,
    npix2nside,
    nside2level,
    reorder,
    ring2double,
    ring2xy,
    to_double_pixelization,
    to_rotated_pixelization,
    xy2xy,
    zonal_average,
)
from earth2grid.healpix.nn import conv2d
from earth2grid.healpix.visualization import pcolormesh

__all__ = [
    "angular_to_global",
    "Compass",
    "conv2d",
    "global_to_angular",
    "global_to_face",
    "Grid",
    "HEALPIX_PAD_XY",
    "local2xy",
    "nest2xy",
    "npix2level",
    "npix2nside",
    "nside2level",
    "pad_backend",
    "pad_with_dim",
    "pad",
    "PaddingBackends",
    "pcolormesh",
    "PixelOrder",
    "reorder",
    "ring2double",
    "ring2xy",
    "to_double_pixelization",
    "to_rotated_pixelization",
    "XY",
    "xy2xy",
    "zonal_average",
]
