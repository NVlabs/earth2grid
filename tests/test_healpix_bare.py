import numpy
import numpy as np
import pytest
import torch

import earth2grid.healpix_bare


def test_ring2nest():
    n = 8
    i = torch.arange(n * n * 12)

    j = earth2grid.healpix_bare.ring2nest(n, i)
    i_round = earth2grid.healpix_bare.nest2ring(n, j)
    numpy.testing.assert_array_equal(i, i_round)


@pytest.mark.parametrize("nest", [True, False])
@pytest.mark.parametrize("lonlat", [True, False])
def test_pix2ang(nest, lonlat, regtest):
    n = 2
    i = torch.arange(n * n * 12)

    x, y = earth2grid.healpix_bare.pix2ang(n, i, nest=nest, lonlat=lonlat)
    print("x", file=regtest)
    np.savetxt(regtest, x, fmt="%.5e")

    print("y", file=regtest)
    np.savetxt(regtest, y, fmt="%.5e")


def savetxt(file, array):
    np.savetxt(file, array, fmt="%.5e")


def test_hpc2loc(regtest):
    x = torch.tensor([0.0]).double()
    y = torch.tensor([0.0]).double()
    f = torch.tensor([0])

    loc = earth2grid.healpix_bare.hpc2loc(x, y, f)
    vec = earth2grid.healpix_bare.loc2vec(loc)
    for array in vec:
        savetxt(regtest, array)


def test_boundaries(regtest):
    ipix = torch.tensor([0])
    boundaries = earth2grid.healpix_bare.corners(1, ipix, False)
    assert not torch.any(torch.isnan(boundaries)), boundaries
    assert boundaries.shape == (1, 3, 4)
    savetxt(regtest, boundaries.flatten())


def test_get_interp_weights_vector():
    lon = torch.tensor([23, 84, -23]).float()
    lat = torch.tensor([0, 12, 67]).float()
    pix, weights = earth2grid.healpix_bare.get_interp_weights(8, lon, lat)
    assert pix.device == lon.device
    assert pix.shape == (4, 3)
    assert weights.shape == (4, 3)


def test_get_interp_weights_vector_interp_y():
    nside = 16
    inpix = torch.tensor([0, 1, 5, 6])

    lon, lat = earth2grid.healpix_bare.pix2ang(nside, inpix, lonlat=True)
    ay = 0.8

    lonc = (lon[0] + lon[1]) / 2
    latc = lat[0] * ay + lat[3] * (1 - ay)

    pix, weights = earth2grid.healpix_bare.get_interp_weights(nside, lonc.unsqueeze(0), latc.unsqueeze(0))

    assert torch.all(pix == inpix[:, None])
    expected_weights = torch.tensor([ay / 2, ay / 2, (1 - ay) / 2, (1 - ay) / 2]).double()[:, None]
    assert torch.allclose(weights, expected_weights)
