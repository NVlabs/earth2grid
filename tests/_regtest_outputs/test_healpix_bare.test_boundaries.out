7.07107e-01
4.56399e-17
0.00000e+00
7.45356e-01
7.07107e-01
7.45356e-01
0.00000e+00
0.00000e+00
0.00000e+00
6.66667e-01
1.00000e+00
6.66667e-01
